### PR TITLE
Integration with 3Dconnexion input devices

### PR DESCRIPTION
**Pull request by pskowronskiTDx**
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

The contribution was created in whole or in part by me and I
have the right to submit it under the open source license
indicated in the file.

I understand and agree that this project and the contribution
are public and that a record of the contribution (including all
personal information I submit with it, including my sign-off) is
maintained indefinitely and may be redistributed consistent with
this project or the open source license(s) involved.

### CODE REVIEW

File: avogadro/menubuilder.h
Comment (ghutchis) [inline avogadro/menubuilder.h:1]: If it's useful API, just add it, not `#ifdef`

File: avogadro/icons/3dx_pivot.png
Comment (ghutchis) [inline avogadro/icons/3dx_pivot.png:1]: This seems kind of small. Does this work okay on a high-resolution / Retina screen?

**Review by ghutchis — CHANGES_REQUESTED** [review]
These are mostly minor, since the main aspect is your code itself.

**Comment by ghutchis** [comment]
@pskowronskiTDx - comments on the requested changes? I'd certainly love to merge these.

File: avogadro/icons/3dx_pivot.png
Comment (pskowronskiTDx) [reply avogadro/icons/3dx_pivot.png:1]: Yes, this icon is meant to be so small.

File: avogadro/tdxcontroller.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
+/******************************************************************************
+  This source file is part of the Avogadro project.
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (pskowronskiTDx) [reply avogadro/tdxcontroller.h:6]: Of course, corrected.

File: thirdparty/3DConnexion/README.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+This directory contains part of the the 3Dconnexion 3DxWare_SDK v4
+The full sdk can be obtained from https://3dconnexion.com/software-developer-program/
```
Comment (ghutchis) [inline thirdparty/3DConnexion/README.txt:2]: Is there a good way to keep this up-to-date (e.g., when v5 comes out?)

File: avogadro/tdxcontroller.cpp [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+}
+
+long Avogadro::TDxController::GetModelExtents(navlib::box_t &extents) const
+{
+  if (m_pGLRenderer == nullptr)
+    return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+
+  std::vector<bool> flags;
+
+  m_pGLRenderer->scene().getBoundingBox(extents.min.x,
```
Comment (ghutchis) [inline avogadro/tdxcontroller.cpp:341]: See, I think you can get this from the molecule here and avoid having to save all those spheres in the render pipeline.

Just implement `core/molecule.*:getBoundingBox()` instead.

File: avogadro/tdxcontroller.cpp [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+{
+  if (m_pGLRenderer == nullptr || *m_ppMolecule == nullptr)
+    return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+
+  std::vector<bool> flags((*m_ppMolecule)->atomCount());
+
+  for (uint32_t i = 0u; i < (*m_ppMolecule)->atomCount(); i++)
+    flags[i] = (*m_ppMolecule)->atomSelected(i);
+
+  m_pGLRenderer->scene().getBoundingBox(extents.min.x,
```
Comment (ghutchis) [inline avogadro/tdxcontroller.cpp:362]: Ditto. You're already going through the atoms in the loop above.

```
for (Index i = 0; i < molecule->atomCount(); ++i) {
   if (molecule->atomSelected(i) {
     // calculate the bounding box directly
   }
}
```

File: avogadro/tdxcontroller.cpp [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+
+long Avogadro::TDxController::GetSelectionTransform(
+  navlib::matrix_t& transform) const
+{
+  return navlib::make_result_code(navlib::navlib_errc::no_data_available);
+}
+
+long Avogadro::TDxController::GetIsSelectionEmpty(navlib::bool_t &empty) const
+{
+  if (*m_ppMolecule == nullptr)
```
Comment (ghutchis) [inline avogadro/tdxcontroller.cpp:381]: Use `Molecule:isSelectionEmpty()`

File: avogadro/tdxcontroller.h [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+ * This class is responsible for handling the TDx navigation in Avogadro2.
+ */
+class TDxController : private TDx::SpaceMouse::Navigation3D::CNavigation3D, private QObject
+{
+
+
+public:
+  TDxController(MainWindow* const mainWindow,
+                Avogadro::QtOpenGL::GLWidget* const pGLWidget,
+                QtGui::Molecule** ppMolecule);
```
Comment (ghutchis) [inline avogadro/tdxcontroller.h:52]: I don't understand the double pointer. Is there a good reason for this?

**Review by ghutchis — CHANGES_REQUESTED** [review]

**Comment by ghutchis** [comment]
Looks pretty close - sorry for the delay. If you could run clang-format on your changed files in `avogadro` (i.e. outside of `thirdparty`) it would be appreciated.

File: avogadro/tdxcontroller.h (hunk 6 above)
Comment (pskowronskiTDx) [reply avogadro/tdxcontroller.h:52]: The double pointer is used to keep track of the actual pointer to active molecule. Then there is no need to update the pointer within the navigation object  each time the active molecule is switched in the explorer. Is there a different approach I should folllow?

File: thirdparty/3DConnexion/README.txt (hunk 2 above)
Comment (pskowronskiTDx) [reply thirdparty/3DConnexion/README.txt:2]: No, as we are shipping some SDK's headers with the solution they will need to be updated manually when a new release comes out.

File: avogadro/tdxcontroller.cpp (hunk 3 above)
Comment (pskowronskiTDx) [reply avogadro/tdxcontroller.cpp:341]: Done as suggested.

**Comment by pskowronskiTDx** [comment]
Thank you for your sugesstions, I have introduced some changes. The double pointer for Molecule object was used to keep track on the active molecule (selected in the explorer). However, I have reworked it by introducing updateMolecule method, connected to the moleculeChanged signal. The pointer to the active molecule updates automatically now, without referencing "pp" anymore.

**Comment by ghutchis** [comment]
Assuming this builds cleanly, I'll merge this as well. Thanks!

**Comment by ghutchis** [comment]
Sorry for the delay - I thought I had previously merged this before leaving for a conference. I'll update the clang-format myself in a new pull.

Thanks!

**Comment by pskowronskiTDx** [comment]
No problem, I am happy to see the changes finally merged. Hope the integration will be useful. Thank you as well!

**Pull request merged**